### ISSUE TITLE
Expo SDK 51: AirGoogleMaps dir must be added to your xCode project to support GoogleMaps on iOS

### ISSUE THREAD

**Issue by robinclaes** [react-native-maps/react-native-maps#5049]
### Summary

After upgrading to Expo SDK 51, and using Expo Go, we receive the following error. We did not have this error with SDK 50.

![WhatsApp Image 2024-05-09 at 09 40 33](https://github.com/react-native-maps/react-native-maps/assets/11699918/7a52554b-a446-4acf-82bf-1dd6b7415c22)

### Reproducible sample code

```TSX
import MapView, { PROVIDER_GOOGLE } from 'react-native-maps'

<MapView
    provider={PROVIDER_GOOGLE}
/>
```

### Steps to reproduce

Start an Expo SDK 51 project, with the managed flow.
Run the app via Expo Go
Use provider `google`

### Expected result

The map renders

### Actual result

We get an error, saying a native dependency is missing.
But since this is Expo managed, it should be fixed with them.

### React Native Maps Version

1.14.0

### What platforms are you seeing the problem on?

iOS (Google Maps)

### React Native Version

0.74.1

### What version of Expo are you using?

SDK 48

### Device(s)

iPhone SE 3 (simulator), iPhone 14 Pro Max (device)

**Comment by Tadeo-Guerstein** [comment]
I have the same issue, is there any way to fix it?

**Comment by flindros** [comment]
I have the same issue. 

Default provider works fine on my apple device but renders Apple Maps

**Comment by janis-ax** [comment]
+1

**Comment by lorissikora** [comment]
Same issue here +1

**Comment by estellerosemary** [comment]
Facing the same issue here.

**Comment by karimcambridge** [comment]
Same issue here!!!

**Comment by samugit83** [comment]
I'm facing the same issue. Version 50 worked perfectly..... and iOS is forcing an update to version 51.

**Comment by elisabeth0bangoura** [comment]
Same here

**Comment by andywalt** [comment]
Me too. Glad to know I'm not alone!

**Comment by brentvatne** [comment]
Hi there! **Google Maps is no longer supported in Expo Go for iOS.** You can use Apple Maps in Expo Go instead (remove the prop `provider={PROVIDER_GOOGLE}` from your `<MapView /> ` components) or switch to a development build (recommended). If you have already configured Google Maps for your production releases, no additional configuration will be necessary. [Learn how to set up a development build](https://docs.expo.dev/get-started/set-up-your-environment/?mode=development-build).

A few things to note about Expo Go:

- Expo Go is intended to be a sandbox environment, and we recommend using development builds for any real app. The ["Set up your environment" guide](https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build) will help you get set up with a development build. We also have a full [EAS tutorial](https://docs.expo.dev/tutorial/eas/introduction/) if you want to learn more, and [a concise video course](https://egghead.io/courses/build-and-deploy-react-native-apps-with-expo-eas-85ab521e) that you can pick and choose lessons from as they are relevant to your use case.
- Development builds give you full control over the native runtime (you can write native code or include any library you like), they are much closer to your production environment, and you can update SDK versions at whichever pace works for you.
- You can still install Expo Go for any SDK version on Android emulators, Android devices, and iOS simulators. We've made the process for doing this as easy as possible on https://expo.dev/go.

**Comment by elisabeth0bangoura** [comment]
I guess I have to switch to native because that was the only reason I used Expo... 😒

**Comment by brentvatne** [comment]
> @elisabeth0bangoura - I guess I have to switch to native because that was the only reason I used Expo... 😒

I'm not sure I understand your comment :) Expo is much more than the "Expo Go" app - the sandbox app for experimenting and learning. I encourage you to take a look at [development builds](https://docs.expo.dev/develop/development-builds/introduction/) - which are like a version of Expo Go that you have full control over.

Ultimately, if you prefer to use other tooling, you should do what is best for you!

**Comment by elisabeth0bangoura** [comment]
> > @elisabeth0bangoura - I guess I have to switch to native because that was the only reason I used Expo... 😒
> 
> I'm not sure I understand your comment :) Expo is much more than the "Expo Go" app - the sandbox app for experimenting and learning. I encourage you to take a look at [development builds](https://docs.expo.dev/develop/development-builds/introduction/) - which are like a version of Expo Go that you have full control over.
> 
> Ultimately, if you prefer to use other tooling, you should do what is best for you!

Many developers were initially attracted to Expo for its simplicity, especially for testing features like maps quickly. However, recent changes may have made the development process feel more complex.

Will Google Maps work in Build development?

**Comment by karimcambridge** [comment]
That's crazy. Picture using development build on simulator on your laptop then expo go on your phone.

Lol this just made things way more complicated. Damn. Thanks for the update tho!

**Comment by elisabeth0bangoura** [comment]
> That's crazy. Picture using development build on simulator on your laptop then expo go on your phone.
> 
> Lol this just made things way more complicated. Damn. Thanks for the update tho!

True!!!!
But I tested it on Build Development and Google Maps on IOS is working there. But yeah they made it so complicated now and with all these files...

**Comment by karimcambridge** [comment]
Yeah because for me, I’m doing like a multi app setting, so I need 2 apps open at the same time. Just was easier that way, but meh, can simple just do a constants check if in `__DEV__` to use not use google maps provider and just have production / dev builds use google.

<3

**Comment by brentvatne** [comment]
> @karimcambridge: That's crazy. Picture using development build on simulator on your laptop then expo go on your phone.
> 
> Lol this just made things way more complicated. Damn. Thanks for the update tho!

you can create a development build for your phone, most developers using them do this! https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build

**Comment by lluisandreu** [comment]
Why is this issue closed? I don't use Expo Go, I ejected Expo a long time ago and I have the same issue. If there is a way to fix it on xCode please share it.

**Comment by brentvatne** [comment]
@lluisandreu - because the issue says in the first line of the summary:

> After upgrading to Expo SDK 51, and using Expo Go, we receive the following error.

if you are experiencing an issue related to this in another context, you should create a new issue and provide all of the relevant information

**Comment by robinclaes** [comment]
> Hi there! **Google Maps is no longer supported in Expo Go for iOS.** You can use Apple Maps in Expo Go instead (remove the prop `provider={PROVIDER_GOOGLE}` from your `<MapView /> ` components) or switch to a development build (recommended). If you have already configured Google Maps for your production releases, no additional configuration will be necessary. [Learn how to set up a development build](https://docs.expo.dev/get-started/set-up-your-environment/?mode=development-build).
> 
> A few things to note about Expo Go:
> 
> * Expo Go is intended to be a sandbox environment, and we recommend using development builds for any real app. The ["Set up your environment" guide](https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build) will help you get set up with a development build. We also have a full [EAS tutorial](https://docs.expo.dev/tutorial/eas/introduction/) if you want to learn more, and [a concise video course](https://egghead.io/courses/build-and-deploy-react-native-apps-with-expo-eas-85ab521e) that you can pick and choose lessons from as they are relevant to your use case.
> * Development builds give you full control over the native runtime (you can write native code or include any library you like), they are much closer to your production environment, and you can update SDK versions at whichever pace works for you.
> * You can still install Expo Go for any SDK version on Android emulators, Android devices, and iOS simulators. We've made the process for doing this as easy as possible on https://expo.dev/go.

@brentvatne thanks, anywhere where we can read up on the decision to remove this from Expo Go? Interested to know the Why.

**Comment by walker007** [comment]
I'm working with two apps using the react-native-maps package; both are with SDK 51, but the first app runs the MapView on iOS, but the second app doesn't.

**Comment by brentvatne** [comment]
@walker007 - that is not a problem. you can learn how to create a [development build](https://docs.expo.dev/develop/development-builds/introduction/) here: https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build

learn more about developing an app with expo here: https://docs.expo.dev/workflow/overview/

**Comment by walker007** [comment]
> is frustrating because it's unexpecte



> @walker007 - that is not a problem. you can learn how to create a [development build](https://docs.expo.dev/develop/development-builds/introduction/) here: https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build
> 
> learn more about developing an app with expo here: https://docs.expo.dev/workflow/overview/

ok, I'll try it, thank you

**Comment by SimonVillage** [comment]
@brentvatne why is nothing about this mentioned here: https://docs.expo.dev/versions/latest/sdk/map-view/#deploy-app-with-google-maps

**Comment by TinasheMuchabaiwa** [comment]
> https://docs.expo.dev/get-started/set-up-your-environment/?platform=ios&device=physical&mode=development-build

Some of us use old Macs that can't update to macOS 14. We heavily relied on expo Go for development. You could've just displayed warnings or disclaimers about potential cases where behavior won't immediately transfer. Honestly, who wouldn't know that they need to use their own API credentials for google maps?

**Comment by WahidN** [comment]
I have a development build + Google Maps API key but I am still getting this error

**Comment by GaWr26** [comment]
I also have a development build with API key and get the error

**Comment by nwachukwuoparah** [comment]
I also have a development build with API key and get the error

**Comment by andriitaran** [comment]
interestingly, Expo Go + Google Maps works fine on a virtual Android device, but doesn't work on an iOS simulator

**Comment by salah-ghanim** [comment]
thanks @brentvatne for taking the time to clarify what happened with the new Expo go release.

to everyone commenting here, this issue is not being tracked because the comments fall into two categories
* complaining about the decision by the expo team (obviously out of scope)
* describing a problem with expo itself (not expo go) with google maps (please submit a new issue).

**Comment by jmejiamu** [comment]
> @brentvatne why is nothing about this mentioned here: https://docs.expo.dev/versions/latest/sdk/map-view

They should add that in the docs

**Comment by jvegax** [comment]
Just read de docs and follow the steps
https://docs.expo.dev/versions/latest/sdk/map-view/

**Comment by tsoisauce** [comment]
> because it used our Google API credentials, but when you create a development build or production build, it would no longer work and you would need to then create your own Google Maps credentials configure it. this is frustrating because it's unexpected and at this point you'll think you've already built support for maps, you may not have budgeted the time in for your release, or maybe you'll just give up and go back to Expo Go. the state we're moving towards is that, whenever possible, something that runs in Expo Go should run without changes in a development build.

Totally understand and much appreciated that this was so painless using your built in Google Map credentials for Expo Go....but if the reason was the credentials in a production build or development build did not transfer over, wouldn't that just be an easy fix?  like updating your app configs/environment variables in `app.json`?  

To me I understand the reasoning but the solution seems really simple.  The ability to play with Google Maps in sandbox out the box outweighs what would just be a credentials update when switching over to development mode.

**Comment by tsoisauce** [comment]
> interestingly, Expo Go + Google Maps works fine on a virtual Android device, but doesn't work on an iOS simulator

I believe its because it defaults to the device provided map.  on iOS it would be Apple Maps and on Android OS it would be Google Map.  Unless you explicitly define the provider in your MapView

**Comment by djnewage** [comment]
adding this just incase anyone needs the help i only ran into this issue running  "expo run:ios", but not "start": "expo start","expo start",

**Comment by EthanHealy01** [comment]
You guys should reconsider your choice to remove Google Maps for IOS. Apple maps cannot be styled and with google maps not working. I don't think theres a way to theme the app with the rest of the app. Light mode maps/dark mode maps etc.

**Comment by jmejiamu** [comment]
> You guys should reconsider your choice to remove Google Maps for IOS. Apple maps cannot be styled and with google maps not working. I don't think theres a way to theme the app with the rest of the app. Light mode maps/dark mode maps etc.

I was able to use Google maps. But I had to` npx expo prebuild` then I just followed the docs as if I were using a RN CLI.

**Comment by EthanHealy01** [comment]
> > You guys should reconsider your choice to remove Google Maps for IOS. Apple maps cannot be styled and with google maps not working. I don't think theres a way to theme the app with the rest of the app. Light mode maps/dark mode maps etc.
> 
> I was able to use Google maps. But I had to` npx expo prebuild` then I just followed the docs as if I were using a RN CLI.

Can you provide more info? Still getting the same error. You on Expo v 51?

**Comment by jmejiamu** [comment]
> > > You guys should reconsider your choice to remove Google Maps for IOS. Apple maps cannot be styled and with google maps not working. I don't think theres a way to theme the app with the rest of the app. Light mode maps/dark mode maps etc.
> > 
> > 
> > I was able to use Google maps. But I had to` npx expo prebuild` then I just followed the docs as if I were using a RN CLI.
> 
> Can you provide more info? Still getting the same error. You on Expo v 51?

Once you run  `npx expo prebuild`, it will create the folders for both Android and iOS then just follow this doc - https://github.com/react-native-maps/react-native-maps/blob/master/docs/installation.md

**Comment by EthanHealy01** [comment]
@jmejiamu Thank you ! 🚀

**Comment by WayneJWZLemon** [comment]
I agree with @jmejiamu—running Google Maps with react-native-maps in Expo command no longer works as expected. To resolve this, I ran:
```
npx expo prebuild
```
Next, I opened ios/MyProject.xcworkspace in Xcode, navigated to the AppDelegate.mm file, and added my Google Maps API key following the instructions in the [react-native-maps installation guide](https://github.com/react-native-maps/react-native-maps/blob/master/docs/installation.md).

After launching the app from Xcode to the simulator, Google Maps rendered correctly.

**Comment by 6ixNugget** [comment]
> I agree with @jmejiamu—running Google Maps with react-native-maps in Expo command no longer works as expected. To resolve this, I ran:
> 
> ```
> npx expo prebuild
> ```
> 
> Next, I opened ios/MyProject.xcworkspace in Xcode, navigated to the AppDelegate.mm file, and added my Google Maps API key following the instructions in the [react-native-maps installation guide](https://github.com/react-native-maps/react-native-maps/blob/master/docs/installation.md).
> 
> After launching the app from Xcode to the simulator, Google Maps rendered correctly.

I'm also directly using development build, not Expo Go, but got hit with the same error. @WayneJWZLemon's fix worked for me. To further add onto the description:

1. After `npx expo prebuild` and adding the necessary lines mentioned in the [guide](https://github.com/react-native-maps/react-native-maps/blob/master/docs/installation.md), you could be prompted with the error `'GoogleMaps/GoogleMaps.h' file not found`. To fix this, go to your Podfile (in the .iOS folder created by prebuild) and add   `pod 'GoogleMaps'`. 
2. Combining with the instructions mentioned in the guide, what's actually needed is 
```
  # React Native Maps dependencies
  rn_maps_path = '../node_modules/react-native-maps'
  pod 'react-native-google-maps', :path => rn_maps_path
  pod 'GoogleMaps'
  pod 'Google-Maps-iOS-Utils'
```
and it should be added before the function call `use_native_modules`

So in my case, the code looks like 

```
target 'MyProject' do
  use_expo_modules!
  
  # React Native Maps dependencies
  rn_maps_path = '../node_modules/react-native-maps'
  pod 'react-native-google-maps', :path => rn_maps_path
  pod 'GoogleMaps'
  pod 'Google-Maps-iOS-Utils'

  config = use_native_modules!
  
  use_frameworks! :linkage => podfile_properties['ios.useFrameworks'].to_sym if podfile_properties['ios.useFrameworks']
  use_frameworks! :linkage => ENV['USE_FRAMEWORKS'].to_sym if ENV['USE_FRAMEWORKS']
```

after adding the snippet.

This helped fixing the issue for me.

**Comment by FreddyJohn** [comment]
This is the ENTIRE REASON I stopped using Expo with my Team YEARS AGO! Great to see Expo dev's are still making stupid opinionated decisions that no one agrees with. This is not a multi-platform solution, I should NEVER have to augment native code EVER. Whats the point in using Expo if I need to modify and maintain two code bases?? Please add this back.

**Comment by davidalejandroaguilar** [comment]
For anyone encountering this and uses Expo, I wrote a config plugin https://www.npmjs.com/package/react-native-maps-expo-plugin to do this without you having to manually tweak files:

```bash
npm install react-native-maps-expo-plugin
```

Then add it to your app.json or app.config.js:

```json
{
  "plugins": [
    [
      "react-native-maps-expo-plugin",
      {
        "googleMapsApiKey": "YOUR_GOOGLE_MAPS_API_KEY"
      }
    ]
  ]
}
```

Now you can generate the native code and build:

```bash
# Generate native code
npx expo prebuild

# Build and run on iOS
npx expo run:ios

# Or if using CNG workflow, use --clean to safely regenerate native files
npx expo prebuild --clean
npx expo run:ios
```

<details>
<summary>What is CNG?</summary>

> 
> Continuous Native Generation (CNG) is a process for building an Expo app where your [native projects](https://docs.expo.dev/workflow/overview#android-and-ios-native-projects) are generated on-demand from your app.json and package.json, similar to how your node_modules are generated from your package.json.
> 
> You can add the [native project](https://docs.expo.dev/workflow/overview#android-and-ios-native-projects) directories (android and ios) to your .gitignore and/or delete the project at any time, then re-generate them from the Expo app config with npx expo prebuild whenever required. You might never even run prebuild on your own development machine if you use a cloud-based development workflow.
> 
> Using CNG can make upgrading to new versions of React Native much easier. It can simplify project maintenance and facilitate setting up complex features such as [App Clips](https://github.com/bndkt/react-native-app-clip), [share extensions](https://github.com/timedtext/expo-config-plugin-ios-share-extension), and [push notifications](https://github.com/OneSignal/onesignal-expo-plugin). This is all made possible with [config plugins](https://docs.expo.dev/config-plugins/introduction). Learn more about [CNG](https://docs.expo.dev/workflow/continuous-native-generation).

</details>

Why this is better than tweaking the generated native files yourself:

Here's Expo documentation on why tweaking the files yourself is not a good idea: [https://docs.expo.dev/workflow/overview/#what-if-i-want-to-edit-the](https://docs.expo.dev/workflow/overview/#what-if-i-want-to-edit-the).

In summary, the official react-native-maps installation requires manual modifications to native iOS files that get wiped out when using CNG (`npx expo prebuild --clean`). This plugin automatically handles:

- Adding Google Maps pod to Podfile
- Adding location permission to Info.plist
- Initializing Google Maps in AppDelegate

In other words, it will do the `react-native-maps` installation instructions for you every time, so you can keep enjoying CNG!

<img width="378" alt="image" src="https://github.com/user-attachments/assets/aed05e3b-5910-4081-ac3b-603950b1443e" />

**Comment by MarcHbb** [comment]
> ```shell
> npx expo prebuild
> ```

Hello ! I'm using expo managed workflow and tried your solution. It almost worked. But I don't why ? 

Before using this plugin, the map didn't show unless I set `PROVIDER_DEFAULT` with the `AirGoogleMaps` error
Now I don't have the error anymore, but I have a blank white map

```
import { SCREEN_HEIGHT, SCREEN_WIDTH } from "@/src/constants";
import { StyleSheet } from "react-native";
import MapView from "react-native-maps";
import { PROVIDER_DEFAULT, PROVIDER_GOOGLE } from "react-native-maps";

export default function MapTab() {
	return (
		<MapView
			testID="map-view"
			loadingEnabled
			provider={PROVIDER_GOOGLE}
			initialRegion={{
				latitude: 37.78825,
				longitude: -122.4324,
				latitudeDelta: 0.0922,
				longitudeDelta: 0.0421,
			}}
			style={styles.map}
			showsMyLocationButton={false}
		/>
	);
}

const styles = StyleSheet.create({
	map: {
		height: SCREEN_HEIGHT,
		width: SCREEN_WIDTH,
	},
});
```

**Comment by amoralesdesign** [comment]
After following all your steps, I get this error after doing `npx expo prebuild` and then `npx expo run:ios`:
```
(NOBRIDGE) ERROR  react-native-maps: AirGoogleMaps dir must be added to your xCode project to support GoogleMaps on iOS. [Component Stack]
 (NOBRIDGE) ERROR  Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

Check the render method of `MapView`.
    in Unknown (created by MapView)
    in MapView (created by EventMap)
    in RCTView (created by View)
    in View (created by EventMap)
   ....
```

**Comment by DanielOnramp** [comment]
> For anyone encountering this and uses Expo, I wrote a config plugin https://www.npmjs.com/package/react-native-maps-expo-plugin to do this without you having to manually tweak files:
> 
> npm install react-native-maps-expo-plugin
> Then add it to your app.json or app.config.js:
> 
> {
>   "plugins": [
>     [
>       "react-native-maps-expo-plugin",
>       {
>         "googleMapsApiKey": "YOUR_GOOGLE_MAPS_API_KEY"
>       }
>     ]
>   ]
> }
> Now you can generate the native code and build:
> 
> # Generate native code
> npx expo prebuild
> 
> # Build and run on iOS
> npx expo run:ios
> 
> # Or if using CNG workflow, use --clean to safely regenerate native files
> npx expo prebuild --clean
> npx expo run:ios
> What is CNG?
> Why this is better than tweaking the generated native files yourself:
> 
> Here's Expo documentation on why tweaking the files yourself is not a good idea: https://docs.expo.dev/workflow/overview/#what-if-i-want-to-edit-the.
> 
> In summary, the official react-native-maps installation requires manual modifications to native iOS files that get wiped out when using CNG (`npx expo prebuild --clean`). This plugin automatically handles:
> 
> * Adding Google Maps pod to Podfile
> * Adding location permission to Info.plist
> * Initializing Google Maps in AppDelegate
> 
> In other words, it will do the `react-native-maps` installation instructions for you every time, so you can keep enjoying CNG!
> 
> <img alt="image" width="378" src="https://private-user-images.githubusercontent.com/15317732/398228454-aed05e3b-5910-4081-ac3b-603950b1443e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MDc1MDksIm5iZiI6MTc0NTUwNzIwOSwicGF0aCI6Ii8xNTMxNzczMi8zOTgyMjg0NTQtYWVkMDVlM2ItNTkxMC00MDgxLWFjM2ItNjAzOTUwYjE0NDNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI0VDE1MDY0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3YTBhMGQ2NTgxZTNiYmY1MjdmZWI3M2YwODFhODlhN2EzNjU2ZmEyNTBkMDZkODBiNDkyN2U5MjAxODFlZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-mMcQrlRJYu98qaz_xyNdVwxIxN1fjiq7fP4Cq3OVO4">

HI, thanks for this, but any idea why the app crashes and this error appears when using this plugin?

```[CoreFoundation] *** Terminating app due to uncaught exception 'GMSServicesException', reason: 'Google Maps SDK for iOS must be initialized via [GMSServices provideAPIKey:...] prior to use'
*** First throw call stack:```
